prov/gni/src/gnix_mbox_allocator.c

```diff
@@ -0,0 +1,537 @@
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <sys/statfs.h>
+#include <sys/mman.h>
+
+#include "gnix_mbox_allocator.h"
+#include "gnix_nic.h"
+
+static int __find_huge_page(size_t page_size, char **directory)
+{
+	int bytes_read = -FI_EINVAL;
+	struct statfs pg_size;
+	size_t size = 1024;
+	char *line, *field;
+	FILE *fd;
+
+	if (!directory || !page_size) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid page_size or directory provided.\n");
+		bytes_read = -FI_EINVAL;
+		goto err;
+	}
+
+	fd = fopen("/proc/mounts", "r");
+	if (!fd) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Error opening /proc/mounts: %s\n",
+			  strerror(errno));
+		bytes_read = -FI_EIO;
+		goto err;
+	}
+
+	line = malloc(size);
+	if (!line) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Error allocating line: %s\n",
+			  strerror(errno));
+		bytes_read = -FI_ENOMEM;
+		goto err1;
+	}
+
+	/*
+	 * Look for entries with substring hugetlbfs in /proc/mounts.
+	 */
+	while (getline(&line, &size, fd) != -1) {
+		if (strstr(line, "hugetlbfs")) {
+			field = strtok(line, " ");
+			field = strtok(NULL, " ");
+
+			/*
+			 * Check if page is the correct size.
+			 */
+			if (statfs(field, &pg_size) == 0) {
+				if (pg_size.f_bsize == page_size) {
+					bytes_read = FI_SUCCESS;
+					break;
+				}
+			}
+		}
+	}
+
+	if (bytes_read == FI_SUCCESS)
+		*directory = strdup(field);
+
+	free(line);
+err1:
+	fclose(fd);
+err:
+	return bytes_read;
+}
+
+static int __generate_file_name(size_t page_size, char **filename)
+{
+	static const char basename[] = "gnix_map";
+	char *full_filename = NULL;
+	static int file_id;
+	char *huge_page = NULL;
+	int size;
+	int ret;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	if (!filename) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "filename pointer is NULL.\n");
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	ret = __find_huge_page(page_size, &huge_page);
+	if (ret < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Find huge page returned negative condition.\n");
+		goto err;
+	}
+
+	size = snprintf(NULL, 0, "%s/%s%d", huge_page, basename, file_id);
+	if (size < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error while gathering size for snprintf: %s\n",
+			  strerror(errno));
+		goto err1;
+	}
+
+	full_filename = malloc(size + 1);
+	if (!full_filename) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating full_filename: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err1;
+	}
+
+	sprintf(full_filename, "%s/%s%d", huge_page, basename, file_id);
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL, "Generated filename: %s\n", full_filename);
+
+	*filename = full_filename;
+
+	file_id++;
```

Comment:
Sure. I actually should probably go do a round of updates and use strtok_r, strerror_r, and update file_id to be an atomic counter.